python/foglamp/services/core/snapshot.py

```diff
@@ -0,0 +1,116 @@
+# -*- coding: utf-8 -*-
+
+# FOGLAMP_BEGIN
+# See: http://foglamp.readthedocs.io/
+# FOGLAMP_END
+
+""" Provides utility functions to take snapshot of plugins"""
+
+import os
+from os.path import basename
+import glob
+import json
+import tarfile
+import fnmatch
+import time
+import subprocess
+
+from foglamp.common import logger
+from foglamp.common.common import _FOGLAMP_ROOT
+
+
+__author__ = "Amarendra K Sinha"
+__copyright__ = "Copyright (c) 2019 Dianomic Systems"
+__license__ = "Apache 2.0"
+__version__ = "${VERSION}"
+
+_LOGGER = logger.setup(__name__)
+_NO_OF_FILES_TO_RETAIN = 3
+
+
+class SnapshotPluginBuilder:
+
+    _out_file_path = None
+    _interim_file_path = None
+
+    def __init__(self, snapshot_plugin_dir):
+        try:
+            if not os.path.exists(snapshot_plugin_dir):
+                os.makedirs(snapshot_plugin_dir)
+            else:
+                self.check_and_delete_plugins_tar_files(snapshot_plugin_dir)
+
+            self._out_file_path = snapshot_plugin_dir
+            self._interim_file_path = snapshot_plugin_dir
+        except (OSError, Exception) as ex:
+            _LOGGER.error("Error in initializing SnapshotPluginBuilder class: %s ", str(ex))
+            raise RuntimeError(str(ex))
+
+    async def build(self):
+        def reset(tarinfo):
+            tarinfo.uid = tarinfo.gid = 0
+            tarinfo.uname = tarinfo.gname = "root"
+            return tarinfo
+        try:
+            snapshot_id = str(int(time.time()))
+            snapshot_filename = "snapshot-plugin-{}.tar.gz".format(snapshot_id)
+            tar_file_name = "{}/{}".format(self._out_file_path, snapshot_filename)
+            pyz = tarfile.open(tar_file_name, "w:gz")
+            try:
+                pyz.add("{}/python/foglamp/plugins".format(_FOGLAMP_ROOT), recursive=True)
+                # C plugins location is different with "make install" and "make"
+                if _FOGLAMP_ROOT == '/usr/local/foglamp':
+                    pyz.add("{}/plugins".format(_FOGLAMP_ROOT), recursive=True, filter=reset)
+                else:
+                    pyz.add("{}/C/plugins".format(_FOGLAMP_ROOT), recursive=True)
```

Comment:
`C-plugins` location is wrong when make, it is under `FOGLAMP_ROOT/cmake_build/C/plugins/`.

And to avoid different build process check please use `FOGLAMP_ROOT/plugins` symlink directory; as we have already discovered C-plugins from the same location.

Reply:
C plugins snapshot actually means creating snapshot of the C source and NOT the final build artifacts. Hence the reason for selecting FOGLAMP_ROOT/C/plugins as C plugins location. Whenever we will restore a plugin snapshot, we will be required to run make/make install again.

Reply:
As such no source file exists for external plugins ONLY compiled. So for `make` process we need to add below combination
```
FOGLAMP_ROOT/cmake_build/C/plugins/ -- Compiled library
FOGLAMP_ROOT/plugins --- SYMLINK
```